crates/apollo-compiler/tests/introspection_split.rs

```diff
@@ -0,0 +1,382 @@
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_both() {
+    let doc = r#"
+        query TheOperation { value __schema { description } }
+    "#;
+    let expected = expect![[r#"
+        Introspection parts:
+        query TheOperation {
+          __schema {
+            description
+          }
+        }
+
+        Other parts:
+        query TheOperation {
+          value
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_nested() {
+    let doc = r#"
+        query TheOperation { value next { __schema { description } } }
+    "#;
+    let expected = expect![[r#"
+        GraphQLError {
+            message: "Schema introspection field __schema is only supported at the root of a query",
+            locations: [
+                GraphQLLocation {
+                    line: 2,
+                    column: 43,
+                },
+            ],
+            path: [],
+            extensions: {},
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_nested_in_fragment() {
+    let doc = r#"
+        query TheOperation { ...A next { ...A } }
+        fragment A on Fibonacci { value __schema { description } }
+    "#;
+    let expected = expect![[r#"
+        GraphQLError {
+            message: "Schema introspection field __schema is only supported at the root of a query",
+            locations: [
+                GraphQLLocation {
+                    line: 3,
+                    column: 41,
+                },
+            ],
+            path: [],
+            extensions: {},
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_mutation() {
+    let doc = r#"
+        mutation TheOperation { __schema { description } }
+    "#;
+    let expected = expect![[r#"
+        GraphQLError {
+            message: "Schema introspection field __schema is only supported at the root of a query",
+            locations: [
+                GraphQLLocation {
+                    line: 2,
+                    column: 33,
+                },
+            ],
+            path: [],
+            extensions: {},
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_mutation_nested() {
+    let doc = r#"
+        mutation TheOperation { value next { __schema { description } } }
+    "#;
+    let expected = expect![[r#"
+        GraphQLError {
+            message: "Schema introspection field __schema is only supported at the root of a query",
+            locations: [
+                GraphQLLocation {
+                    line: 2,
+                    column: 46,
+                },
+            ],
+            path: [],
+            extensions: {},
+        }
+    "#]];
+    assert_split(doc, expected);
+}
+
+#[test]
+fn test_nested_in_other_operation() {
+    let doc = r#"
+        query TheOperation { value }
+        query OtherOperation { value next { __schema { description } } }
+    "#;
+    let expected = expect!["None"];
```

Comment:
This is not meant to produce an error like with the cases above?

Reply:
No. You’d get an "unsupported" error if you select the other operation, but the document (however weird) is valid per spec.